src/probability_theory/density.lean

```diff
@@ -0,0 +1,374 @@
+begin
+  by_cases hpdf : integrable (λ x, f x * (pdf X ℙ μ x).to_real) μ,
+  { rw [← integral_map hX hf.ae_measurable, map_eq_with_density_pdf X ℙ μ,
+        integral_eq_lintegral_pos_part_sub_lintegral_neg_part hpdf,
+        integral_eq_lintegral_pos_part_sub_lintegral_neg_part,
+        lintegral_with_density_eq_lintegral_mul _ (measurable_pdf X ℙ μ) hf.neg.ennreal_of_real,
+        lintegral_with_density_eq_lintegral_mul _ (measurable_pdf X ℙ μ) hf.ennreal_of_real],
+    { congr' 2,
+      { have : ∀ x, ennreal.of_real (f x * (pdf X ℙ μ x).to_real) =
+          ennreal.of_real (pdf X ℙ μ x).to_real * ennreal.of_real (f x),
+        { intro x,
+          rw [mul_comm, ennreal.of_real_mul ennreal.to_real_nonneg] },
+        simp_rw [this],
+        exact lintegral_congr_ae (filter.eventually_eq.mul
+          (of_real_to_real_ae_eq hX) (ae_eq_refl _)) },
+      { have : ∀ x, ennreal.of_real (- (f x * (pdf X ℙ μ x).to_real)) =
+          ennreal.of_real (pdf X ℙ μ x).to_real * ennreal.of_real (-f x),
+        { intro x,
+          rw [neg_mul_eq_neg_mul, mul_comm, ennreal.of_real_mul ennreal.to_real_nonneg] },
+        simp_rw [this],
+        exact lintegral_congr_ae (filter.eventually_eq.mul
+          (of_real_to_real_ae_eq hX) (ae_eq_refl _)) } },
+    { refine ⟨hf.ae_measurable, _⟩,
+      rw [has_finite_integral, lintegral_with_density_eq_lintegral_mul _
+            (measurable_pdf _ _ _) hf.nnnorm.coe_nnreal_ennreal],
+      have : (λ x, (pdf X ℙ μ * λ x, ↑∥f x∥₊) x) =ᵐ[μ] (λ x, ∥f x * (pdf X ℙ μ x).to_real∥₊),
+      { simp_rw [← smul_eq_mul, nnnorm_smul, ennreal.coe_mul],
+        rw [smul_eq_mul, mul_comm],
+        refine filter.eventually_eq.mul (ae_eq_refl _)
+          (ae_eq_trans (of_real_to_real_ae_eq hX).symm _),
+        convert ae_eq_refl _,
+        ext1 x,
+        exact real.ennnorm_eq_of_real ennreal.to_real_nonneg },
+      rw lintegral_congr_ae this,
+      exact hpdf.2 } },
+  { rw [integral_undef hpdf, integral_undef],
+    rwa ← integrable_iff_integrable_mul_pdf hX hf at hpdf,
+    all_goals { apply_instance } }
+end
+
+/-- A random variable that `has_pdf` is quasi-measure preserving. -/
+lemma to_quasi_measure_preserving (X : α → E) (hX : measurable X) [has_pdf X ℙ μ] :
+  quasi_measure_preserving X ℙ μ :=
+{ measurable := hX,
+  absolutely_continuous :=
+  begin
+    rw map_eq_with_density_pdf X ℙ μ,
+    exact with_density_absolutely_continuous _ _,
+    all_goals { apply_instance }
+  end }
+
+lemma map_absolutely_continuous {X : α → E} (hX : measurable X) [has_pdf X ℙ μ] :
+  map X ℙ ≪ μ :=
+(to_quasi_measure_preserving X hX).absolutely_continuous
+
+end
+
+section real
+
+variables [is_finite_measure ℙ] {X : α → ℝ} (hX : measurable X)
+
+include hX
+
+/-- A real-valued random variable `X` `has_pdf X ℙ λ` (where `λ` is the Lebesgue measure) if and
+only if the push-forward measure of `ℙ` along `X` is absolutely continuous with respect to `λ`. -/
+lemma has_pdf_iff : has_pdf X ℙ ↔ map X ℙ ≪ volume :=
```

Comment:
You can generalize this to measures other than volume, and to spaces beyond R. The following code compiles if you put it right after `map_absolutely_continuous`:
```

/-- TODO: add this where zero is, rename zero to zero_right, add simp attributes -/
lemma mutually_singular.zero_left {μ : measure E} : 0 ⊥ₘ μ :=
mutually_singular.symm mutually_singular.zero

/-- TODO: move where appropriate. -/
lemma have_lebesgue_decomposition_of_has_pdf {X : α → E} [hX' : has_pdf X ℙ μ] :
  (map X ℙ).have_lebesgue_decomposition μ :=
⟨⟨⟨0, pdf X ℙ μ⟩,
  by simp only [zero_add, measurable_pdf X ℙ μ, true_and, mutually_singular.zero_left,
    map_eq_with_density_pdf X ℙ μ] ⟩⟩

/-- TODO: move where appropriate. -/
lemma has_pdf_iff {X : α → E} (hX : measurable X) :
  has_pdf X ℙ μ ↔ (map X ℙ).have_lebesgue_decomposition μ ∧ map X ℙ ≪ μ :=
begin
  split,
  { intro hX',
    exactI ⟨have_lebesgue_decomposition_of_has_pdf, map_absolutely_continuous hX⟩, },
  { rintros ⟨h_decomp, h⟩,
    haveI := h_decomp,
    refine ⟨⟨(measure.map X ℙ).radon_nikodym_deriv μ, measurable_radon_nikodym_deriv _ _, _⟩⟩,
    rwa with_density_radon_nikodym_deriv_eq, }
end
```
In the real case with volume and a finite P, you get the decomposition for free.

Reply:
In this comment and the one below, I am not saying that you should delete the real-and-volume case. I suggest having the general case, then a specialized lemma.

Reply:
Thanks!